Mathlib/NumberTheory/LocalField/Basic.lean

```diff
@@ -0,0 +1,176 @@
+  by_cases hγ : γ = 0
+  · simp [hγ]
+  letI := IsTopologicalAddGroup.toUniformSpace K
+  letI := isUniformAddGroup_of_addCommGroup (G := K)
+  obtain ⟨s, hs, -, hs'⟩ := LocallyCompactSpace.local_compact_nhds (0 : K) .univ Filter.univ_mem
+  obtain ⟨r, hr, hr1, H⟩ :
+      ∃ r', r' ≠ 0 ∧ valuation K r' < 1 ∧ { x | valuation K x ≤ valuation K r' } ⊆ s := by
+    obtain ⟨r, hr, hrs⟩ := (ValuativeTopology.hasBasis_nhds_zero' K).mem_iff.mp hs
+    obtain ⟨r', hr', hr⟩ := Valuation.IsNontrivial.exists_lt_one (v := valuation K)
+    simp only [ne_eq, map_eq_zero] at hr'
+    obtain hr1 | hr1 := lt_or_ge r 1
+    · obtain ⟨r, rfl⟩ := ValuativeRel.valuation_surjective r
+      simp only [ne_eq, map_eq_zero] at hr
+      refine ⟨r ^ 2, by simpa using hr, by simpa [pow_two], fun x hx ↦ hrs ?_⟩
+      simp only [map_pow, Set.mem_setOf_eq] at hx ⊢
+      refine hx.trans_lt (by simpa [pow_two, hr])
+    · refine ⟨r', hr', hr, .trans ?_ hrs⟩
+      intro x hx
+      dsimp at hx ⊢
+      exact hx.trans_lt (hr.trans_le hr1)
+  convert (hs'.of_isClosed_subset (Valued.isClosed_closedBall K _) H).image
+    (Homeomorph.mulLeft₀ (γ / r) (by simp [hr, div_eq_zero_iff, hγ])).continuous using 1
+  refine .trans ?_ (Equiv.image_eq_preimage _ _).symm
+  ext x
+  simp [div_mul_eq_mul_div, div_le_iff₀, v_eq_valuation, hγ, hr]
+
+instance : CompactSpace 𝒪[K] := isCompact_iff_compactSpace.mp (isCompact_closedBall K 1)
+
+instance (K : Type*) [Field K] [ValuativeRel K] [UniformSpace K] [IsUniformAddGroup K]
+    [ValuativeTopology K] : (Valued.v (R := K) (Γ₀ := ValueGroupWithZero K)).Compatible :=
+  inferInstanceAs (valuation K).Compatible
+
+instance : IsDiscreteValuationRing 𝒪[K] :=
+  letI := IsTopologicalAddGroup.toUniformSpace K
+  haveI := isUniformAddGroup_of_addCommGroup (G := K)
+  haveI : CompactSpace (Valued.integer K) := inferInstanceAs (CompactSpace 𝒪[K])
+  Valued.integer.isDiscreteValuationRing_of_compactSpace
+
+/-- The value group of a local field is (uniquely) isomorphic to `ℤᵐ⁰`. -/
+noncomputable
+def valueGroupWithZeroIsoInt : ValueGroupWithZero K ≃*o ℤᵐ⁰ := by
+  apply Nonempty.some
+  letI := IsTopologicalAddGroup.toUniformSpace K
+  haveI := isUniformAddGroup_of_addCommGroup (G := K)
+  obtain ⟨_⟩ := Valued.integer.locallyFiniteOrder_units_mrange_of_isCompact_integer
+    (isCompact_iff_compactSpace.mpr (inferInstanceAs (CompactSpace 𝒪[K])))
+  let e : (MonoidHom.mrange (valuation K)) ≃*o ValueGroupWithZero K :=
+    ⟨.ofBijective (MonoidHom.mrange (valuation K)).subtype ⟨Subtype.val_injective, fun x ↦
+      ⟨⟨x, ValuativeRel.valuation_surjective x⟩, rfl⟩⟩, .rfl⟩
+  have : Nontrivial (ValueGroupWithZero K)ˣ := isNontrivial_iff_nontrivial_units.mp inferInstance
+  have : Nontrivial (↥(MonoidHom.mrange (valuation K)))ˣ :=
+    (Units.map_injective (f := e.symm.toMonoidHom) e.symm.injective).nontrivial
+  exact ⟨e.symm.trans (LocallyFiniteOrder.orderMonoidWithZeroEquiv _)⟩
+
+instance : ValuativeRel.IsDiscrete K where
+  has_maximal_element := by
```

Comment:
How about doing things in prop land before having to make the explicit noncomputable iso?
- installing a LocallyFiniteOrder (ValueGroupWithZero K)^x, from Valued.integer.locallyFiniteOrder_units_mrange_of_isCompact_integer
- installing a MulArchimedean (ValueGroupWithZero K), from the hom you've written
- installing a IsRankLeOne from MulArchimedean, via #26833
- IsDiscrete R by #27213, `IsDiscrete.of_compatible_withZeroMulInt`, which you have because you have (f : ValueGroupWithZero R →*₀o ℤᵐ⁰)` and your compatible v is `(valuation R).comp f`
```

Reply:
I think the iso is fine because it is unique, and the defeq is probably always bad for local fields.

Reply:
I don't understand "installing a IsRankLeOne from MulArchimedean". The linked PR doesn't seem to contain the advertised result.

Reply:
But I am indeed looking for a better way to do all this. As Kenny has pointed out above, we essentially already have rank one during the proof of DVR.

Reply:
I'm not sure I understand this comment, the proof you linked to basically already do the first three points. For the last point, I now have a very short proof (thanks to your previous PR's) in https://github.com/erdOne/mathlib4/pull/1